spring-kafka/src/main/java/org/springframework/kafka/listener/PartitionPausingBackOffManagerFactory.java

```diff
@@ -161,7 +159,15 @@ private TaskExecutor getOrCreateTimingAdjustmentThreadExecutor() {
 		}
 		ThreadPoolTaskExecutor executor = new ThreadPoolTaskExecutor();
 		executor.initialize();
-		super.addApplicationListener((ApplicationListener<ContextClosedEvent>) event -> executor.shutdown());
+		this.taskExecutor = executor;
 		return executor;
 	}
+
+	@Override
+	public void destroy() throws Exception {
+		if (this.taskExecutor != null
```

Comment:
I think this is not this factory responsibility. 
Consider to move logic for the default `executor` directly to that `WakingKafkaConsumerTimingAdjuster`.
Do we really need a thread pool over there for that `adjustTiming()`?
Isn't a single dedicated thread enough?
Kinda `Executors.newSingleThreadExecutor()` ?

Reply:
> I think this is not this factory responsibility.
> Consider to move logic for the default `executor` directly to that `WakingKafkaConsumerTimingAdjuster`.

Sure, I'll take a look into that, thanks.

> Do we really need a thread pool over there for that `adjustTiming()`?
> Isn't a single dedicated thread enough?
> Kinda `Executors.newSingleThreadExecutor()` ?

Currently we adjust the timing for increasing the `back off`'s precision by spawning a thread that will `wakeup` the consumer at the appropriate time, if there's such relevant adjustment to be made. This was created this way to keep it simple - fire and forget, no thread management required.

Since normally these should be short lived tasks, it shouldn't be much of a performance issue having that many threads and the thread pool assures we reuse the threads. Also, the threads spend most of their time sleeping, so there's not that much CPU usage. If we have a single thread with this strategy, it might get blocked by a longer delay and make other delays less precise.

It might be worth looking at something more of a manager with a single thread that goes through values and wakes up the consumers as needed. It's more complex, and might be a busy thread, but probably would be more performant overall, specially if we kill it when not in use.

But for the solution as is, I think we still need the `ThreadPool`.

WDYT?

Thanks

Reply:
I changed the `TaskExecutor` to the `WakingKafkaConsumerTimingAdjuster` and it indeed looks a lot better, thanks!

I thought of extending the `SmartLifecycle` interface in both the `KafkaConsumerBackOffManager` and `KafkaConsumerTimingAdjuster` interfaces instead of `DisposableBean`. While the abstraction itself seemed more appropriate since these are long running beans, it was clumsy to implement that without breaking backwards compatibility, so for now it doesn't seem like it's worth it - please let me know if you think otherwise.

Thanks a lot for your input, and please let me know if there's anything else.